### PR TITLE
overload toOne() and toOneMany() to take custom assertion message

### PR DESCRIPTION
**Pull request by akphi**

### CODE REVIEW

File: legend-pure-m3-core/src/test/java/org/finos/legend/pure/m3/tests/function/base/multiplicity/AbstractTestToOne.java [hunk 1]
```diff
@@ -24,6 +24,8 @@ public void testToOneError()
     {
         assertExpressionRaisesPureException("Cannot cast a collection of size 0 to multiplicity [1]", 3, 13, "[]->toOne()");
         assertExpressionRaisesPureException("Cannot cast a collection of size 3 to multiplicity [1]", 3, 26, "['a', 'b', 'c']->toOne()");
+        assertExpressionRaisesPureException("Something wrong", 3, 13, "[]->toOne('Something wrong')");
```
Comment (kevin-m-knight-gs) [inline legend-pure-m3-core/src/test/java/org/finos/legend/pure/m3/tests/function/base/multiplicity/AbstractTestToOne.java:27]: You should add a test where the new function doesn't raise an exception.

File: legend-pure-m3-core/src/test/java/org/finos/legend/pure/m3/tests/function/base/multiplicity/AbstractTestToOneMany.java [hunk 2]
```diff
@@ -23,6 +23,7 @@ public abstract class AbstractTestToOneMany extends PureExpressionTest
     public void testToOneManyError()
     {
         assertExpressionRaisesPureException("Cannot cast a collection of size 0 to multiplicity [1..*]", 3, 13, "[]->toOneMany(); 1;");
+        assertExpressionRaisesPureException("Something wrong", 3, 13, "[]->toOneMany('Something wrong'); 1;");
```
Comment (kevin-m-knight-gs) [inline legend-pure-m3-core/src/test/java/org/finos/legend/pure/m3/tests/function/base/multiplicity/AbstractTestToOneMany.java:26]: You should add a test where the new function doesn't raise an exception.

File: legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneManyWithAssert.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2020 Goldman Sachs
```
Comment (kevin-m-knight-gs) [inline legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneManyWithAssert.java:1]:
```suggestion
// Copyright 2022 Goldman Sachs
```

File: legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneManyWithAssert.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.pure.runtime.java.compiled.generation.processors.natives.core.multiplicity;
+
+import org.finos.legend.pure.m4.coreinstance.SourceInformation;
+import org.finos.legend.pure.runtime.java.compiled.generation.processors.natives.AbstractNativeFunctionGeneric;
+import org.finos.legend.pure.runtime.java.compiled.generation.processors.support.CompiledSupport;
+
+public class ToOneManyWithAssert extends AbstractNativeFunctionGeneric
```
Comment (kevin-m-knight-gs) [inline legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneManyWithAssert.java:21]: I think ToOneManyWithMessage would be a better name.

File: legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneWithAssert.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2020 Goldman Sachs
```
Comment (kevin-m-knight-gs) [inline legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneWithAssert.java:1]:
```suggestion
// Copyright 2022 Goldman Sachs
```

File: legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneWithAssert.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+// limitations under the License.
+
+package org.finos.legend.pure.runtime.java.compiled.generation.processors.natives.core.multiplicity;
+
+import org.finos.legend.pure.m4.coreinstance.SourceInformation;
+import org.finos.legend.pure.runtime.java.compiled.generation.processors.natives.AbstractNativeFunctionGeneric;
+import org.finos.legend.pure.runtime.java.compiled.generation.processors.support.CompiledSupport;
+
+
+public class ToOneWithAssert extends AbstractNativeFunctionGeneric
```
Comment (kevin-m-knight-gs) [inline legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/natives/core/multiplicity/ToOneWithAssert.java:22]: I think ToOneWithMessage is a better name.

File: legend-pure-runtime-java-engine-interpreted/src/main/java/org/finos/legend/pure/runtime/java/interpreted/natives/core/meta/ToMultiplicity.java [hunk 7]
```diff
@@ -55,7 +57,12 @@ public CoreInstance execute(ListIterable<? extends CoreInstance> params, Stack<M
         ListIterable<? extends CoreInstance> values = Instance.getValueForMetaPropertyToManyResolved(param, M3Properties.values, processorSupport);
         if (!Multiplicity.isValid(returnMultiplicity, values.size()))
         {
-            throw new PureExecutionException(functionExpressionToUseInStack.getSourceInformation(), "Cannot cast a collection of size " + values.size() + " to multiplicity " + Multiplicity.print(returnMultiplicity));
+            String errorMessage = null;
+            if (this.supportAssertion)
```
Comment (kevin-m-knight-gs) [inline legend-pure-runtime-java-engine-interpreted/src/main/java/org/finos/legend/pure/runtime/java/interpreted/natives/core/meta/ToMultiplicity.java:61]: I don't think you need this "supportAssertion" boolean. You can just check the number of parameters to get the information you want.

File: legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/support/CompiledSupport.java [hunk 8]
```diff
@@ -386,11 +386,31 @@ public static int safeHashCode(Object object)
...
+    public static <T> T toOneWithAssert(T object, String message, SourceInformation sourceInformation)
+    {
+        return toOne(object, message, sourceInformation);
+    }
+
+    public static <T> T toOneWithAssert(RichIterable<? extends T> objects, String message, SourceInformation sourceInformation)
+    {
+        return toOne(objects, message, sourceInformation);
+    }
+
```
Comment (kevin-m-knight-gs) [inline legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/support/CompiledSupport.java:398]: I don't think these functions serve much purpose. Just make the toOne functions with the message parameter public. The same comment applies to the toOneManyWithAssert functions.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]
I think "WithMessage" is a better suffix than "WithAssert". Really, toOne and toOneMany always have asserts. The only question is whether a message is passed in.

File: legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/support/CompiledSupport.java (hunk 8 above)
Comment (akphi) [reply legend-pure-runtime-java-engine-compiled/src/main/java/org/finos/legend/pure/runtime/java/compiled/generation/processors/support/CompiledSupport.java]: Makes sense. Thank you for reviewing!

**Review by kevin-m-knight-gs — APPROVED** [review]

**Comment by rafaelbey** [comment]
Love it!  Thanks for this.

**Pull request merged**